protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
+- upon commit, all the transactions remaining in the mempool are re-checked.
+
+**Correctness**
+We now establish that the above algorithm implements the mempool abstraction.
+For invariants INV2a and INV3a, we show that **there exist** such an $\alpha$ and $\beta$, but we do not look at characterizing them precisely.
+Regarding INV4, we shall assume that step (1) occurs infinitely often and not once upon the reception of the transaction.
+This models that a CometBFT client may re-submit a failed transaction forever until it succeeds (if it does).
+
+The refinement mapping from `CListMempool` to the mempool abstraction is as follows:
+Step (1) increments $p.valid[tx]$, which is initially set to 0.
+Conversely, steps (3) and (5) reset $p.valid[tx]$ to 0.
+Transaction $tx$ is added to $p.lcommitted$ when step (2).
+It is removed when (4) takes place.
+Step (6) adds the transaction to the mempool.
+The transaction is removed from the mempool upon (7) and (8).
+
+Consider some correct process $p$.
+Below, we show in order that all the invariants of the mempool are maintained at $p$:  
+
+- INV1.
+Clear from the code base, as the consensus input corresponds to a call to `Execution:CreateProposalBlock` which itself is calling `ClistMempool:ReapMaxBytesMaxGas`, returning a subset of $p.mempool$.
+
+- INV2a.
+If $tx$ is in $p.lcommitted$, it was added due to step (2).
+This step occurs only if the transaction was committed at some height.
+Furthermore, when this happens step (7) is taken.
+It follows, that the transaction is removed from the mempool, as required.
+
+- INV3a.
+Consider some height $h$ for which $tx$ is in the mempool and $p.valid[tx]>0$.
+For any transaction $tx$, this counter is set to $0$ initially.
+Hence, $p.valid[tx]$ was incremented at some height $h' \leq h$.
+Consider the point in time $t'$, where p moves to height $h$.
+If $p.valid[tx]$ equals 0 at that time, then it was incremented at height h.
+From (6), it is necessarily valid at height h, as requied.
+Otherwise, the transaction is rechecked at the current (new) height h with step (8).
+Because $p.valid[tx]$ is not reset, it follows that the application considers it valid at h 
+
+- INV4.
+Consider that a transaction $tx$ enters the mempool at $p$.
+If $tx$ leaves the mempool, step (7) or (8) happens.
+In the former case, $tx$ is committed as required.
+In the latter, $tx$ is re-checked and invalid at some height $h$.
+Re-checking a transaction happens only once per height, say at time $t$ for transaction $tx$ and height $h$.
+When this happens at time $t$ (line 474), the transaction is also removed from the cache (line 477) due to step (3).
+By assumption, step (1) is eventually re-executed for transaction $tx$ after time $t$.
+Because $tx$ is no more in the cache, it is added again after time $t$.
+Now, assume that transaction $tx$ is not invalid infinitely often at process $p$.
```

Comment:
With the new version of the requirement to the app, we would also need to explain the "opposite" assumption here: that a tx eventually converges to be (and stay) invalid

Reply:
In the proof, I had omitted that eventual non-oscillation was necessary to prove INV4. This assumption is now mentionned.